src/address.js

```diff
@@ -6,30 +6,54 @@ var types = require('./types')
 
 function fromBase58Check (address) {
   var payload = bs58check.decode(address)
-  if (payload.length < 21) throw new TypeError(address + ' is too short')
-  if (payload.length > 21) throw new TypeError(address + ' is too long')
 
   var version = payload[0]
   var hash = payload.slice(1)
+  var segWitVersion
+  var segWitPadding
 
-  return { hash: hash, version: version }
+  if (hash.length === 22 || hash.length === 34) {
+    segWitVersion = hash.readUInt8(0)
+    segWitPadding = hash.readUInt8(1)
+    hash = hash.slice(2)
+
+    if (segWitVersion > 16 || segWitPadding !== 0) {
+      throw new Error(address + ' has the length of a segWit address, but is not a valid segWit address')
+    }
+  } else if (hash.length > 20) throw new Error(address + ' is too long')
+  else if (hash.length < 20) throw new Error(address + ' is too short')
+
+  return { hash: hash, version: version, segWitVersion: segWitVersion }
 }
 
 function fromOutputScript (scriptPubKey, network) {
   network = network || networks.bitcoin
 
   if (bscript.isPubKeyHashOutput(scriptPubKey)) return toBase58Check(bscript.compile(scriptPubKey).slice(3, 23), network.pubKeyHash)
   if (bscript.isScriptHashOutput(scriptPubKey)) return toBase58Check(bscript.compile(scriptPubKey).slice(2, 22), network.scriptHash)
+  if (bscript.isSegWitPubKeyHashOutput(scriptPubKey)) return toBase58Check(bscript.compile(scriptPubKey).slice(2, 22), network.segWitPubKeyHash, 0)
+  if (bscript.isSegWitScriptHashOutput(scriptPubKey)) return toBase58Check(bscript.compile(scriptPubKey).slice(2, 34), network.segWitScriptHash, 0)
 
   throw new Error(bscript.toASM(scriptPubKey) + ' has no matching Address')
 }
 
-function toBase58Check (hash, version) {
-  typeforce(types.tuple(types.Hash160bit, types.UInt8), arguments)
+function toBase58Check (hash, version, segWitVersion) {
```

Comment:
They shelved that proposal until later - pending feedback/proposals from wallet developers IIRC